lib.php

```diff
@@ -0,0 +1,153 @@
+
+    // Use "/api" endpoint to get key characteristics of the API such as the server name and the default version.
+    $resource = $apiurl . '/api';
+    $serverinfo = $curl->get($resource);
+
+    // Get curl request info.
+    $info = $curl->get_info();
+    // We don't accept anything other than http_code 200.
+    if (!isset($info['http_code']) || $info['http_code'] != 200) {
+        // If we want the result to be displayed in adamin page with redirect action.
+        if ($redirect) {
+            // Redirect back to admin page with related error message.
+            redirect(new \moodle_url('/admin/settings.php?section=tool_opencast'),
+                get_string('apiurltestfailedlong', 'tool_opencast') , 0, \core\output\notification::NOTIFY_ERROR);
+        }
+
+        // Otherwise, we create the error message text with html_write here and return it.
+        return \html_writer::tag('p', get_string('apiurltestfailedshort', 'tool_opencast'), array('class' => 'alert alert-danger'));
+    }
+
+    //In case given URL is valid, we now check if the username and password is provided.
+    // Get apiusername from admin setting.
+    $apiusername = get_config('tool_opencast', 'apiusername');
+
+    // Get apipassword from admin setting as well.
+    $apipassword = get_config('tool_opencast', 'apipassword');
+
+    // Here we check if credentials are not empty.
+    if (empty($apiusername) || empty($apipassword)) {
+        // If we want the result to be displayed in adamin page with redirect action.
+        if ($redirect) {
+            // Redirect back to admin page with a warning message to inform admin that credentials are not yet provided.
+            redirect(new \moodle_url('/admin/settings.php?section=tool_opencast'),
+                get_string('apiurltestsucceedbutnocredentialslong', 'tool_opencast') , 0, \core\output\notification::NOTIFY_WARNING);
+        }
+    }
+
+    // In case that redirect is false, it means we call this function from connection test tool and it has to return a successs notification as html tag string.
+    if (!$redirect) {
+        return \html_writer::tag('p', get_string('apiurltestsuccessfulshort', 'tool_opencast'), array('class' => 'alert alert-success'));
+    }
+
+    // When we reach here, it means that everything went fine and the function is called from admin page, therefore
+    // we leave it to the admin_settings to handel the rest from here.
+}
+
+/**
+ * Helperfunction to validate given credentials for the Opencast API URL.
+ * It uses curl with Authorization header.
+ *
+ * @param bool $redirect: When true a redirect action to admin page will occure upon failure. Otherwise, an html tag as string will be returned.
+ * 
+ * @return string in case the function is called through connection test tool an html tag as string will be returned.
+ */
+function tool_opencast_test_connection_with_credentials($redirect = true) {
+    // Initialize cURL.
+    $curl = new \curl();
```

Comment:
Why didn't you extend the `api` class instead of reimplementing the curl connection?